src/cargo/core/dependency.rs

```diff
@@ -104,14 +102,32 @@ fn parse_req_with_deprecated(
     req: &str,
     extra: Option<(PackageId, &Config)>,
 ) -> CargoResult<VersionReq> {
-    match VersionReq::parse(req) {
-        Err(ReqParseError::DeprecatedVersionRequirement(requirement)) => {
-            let (inside, config) = match extra {
-                Some(pair) => pair,
-                None => return Err(ReqParseError::DeprecatedVersionRequirement(requirement).into()),
-            };
-            let msg = format!(
-                "\
+    let err = match VersionReq::parse(req) {
+        Ok(req) => return Ok(req),
+        Err(err) => err,
+    };
+
+    let (inside, config) = match extra {
+        Some(pair) => pair,
+        None => return Err(err.into()),
+    };
+
+    let corrected = match req {
+        ".*" => "*",
+        "0.1.0." => "0.1.0",
+        "0.3.1.3" => "0.3.13",
+        "0.2*" => "0.2.*",
+        "*.0" => "*",
```

Comment:
Thanks for keeping the compat here! That being said the warning below says that it will soon become a hard error, and that was added 5 years ago in https://github.com/rust-lang/cargo/pull/3154. I think it's probably fine to just delete all this at this point (and I'm glad you moved this logic to Cargo from the `semver` crate itself!)

Reply:
Good call :stuck_out_tongue:  I've deleted the compatibility code in 396bdd3a89fe4279374096f95be2f0134f62ea83.